Support certificate verification

jonandernovella opened the pull request:
- Endpoint for verification under `/api/certificates/verify/ind_number`
- Checksum and signature verification is performed

pontus committed:
Use suggested action for codeql

Commit:
First draft of coefficient of inbreed calculations

Commit:
Merge branch 'develop' into feature/test-breed-functionality

Commit:
Draft of inbreed popup with recommendation and pedigree network for chosen parents with their common ancestors highlighted

Commit:
Jest test for pedigree inbreed functionality

Commit:
renamed function colourEdges to getConnectingEdges and modified related code to better match its functionality

Commit:
Added more typing and questions regarding these and other design considerations

Commit:
Update filter labels to match previously updated filter functionality

Commit:
Fixed some feedback

Commit:
Fixed typo

Commit:
Added ability to choose grandparents to breed instead of mother/father, semiugly but working draft

Commit:
Button disabled when parents/grandparents not chosen properly

Commit:
Fixed warnings about controlled and uncontrolled Autocompleteand made the states more separate

Commit:
Modified InbreedingRecommendation and related code to have possibility to accept grandParents as well

Commit:
Combined unique and commonAncestors into method uniqueAndCommonNodes, will replace and remove unique and commonAncestors when brain is not Friday mushy and prone to creating errors

Commit:
Adapted parentPedigree to use uniqueAndCommonAncestors

Commit:
only uniqueAndCommonNodes in parentPedigree and update related test

Commit:
First draft testBreedPedgidree which accept both parents and grandparents

Commit:
Made one common testBreedPedigree function regardless of choosing parent or grandparent

Commit:
Minor cleanup

Commit:
Minor cleanup

Commit:
Reorganizing the styling of the form

Commit:
Number of generations shown in pedigree network based on potential offspring as generation 0

Commit:
Added pedigree network configuration possibilities

andkaha committed:
Add test for duplicate numbers

Check for duplicate numbers for both Gotland and Mellerud data sets.
If a duplicate number is found, issue a diagnostic message about this
and don't even attempt to load the data.

This should fix issue #225.

Commit:
Now possible to toggle if common ancestors should be colored or not

Commit:
Minor layout changes related to coloring common ancestors

Commit:
Small visual prepp for demo

Commit:
One threshold for coefficient of inbreed

Commit:
Added button to choose genebank

Commit:
Merge branch 'develop' into feature/test-breed-functionality

Commit:
Improving layout of the information

Commit:
Merge branch 'develop' into feature/test-breed-functionality

Commit:
Draft of coefficient of inbreed for testbreed offspring. Calculates correctly for mother and father, not yet for grandparents

Commit:
Calculation of COI working when choosing grandparents as well

Commit:
Added genebank id to R testbreed function + small cleanup

Commit:
Fix (draft) so it is possible to write input for male grandparent and not just choose from options

Commit:
Created container for all states since there are quite many. Working for male ancestors

Commit:
States container soultion working for both female and male ancestors

Commit:
Made separation between states of individuals and labels and updated related files accordingly

Commit:
offspringCOI calculated from r-api now used instead of dummy value

Commit:
pedigree now functioning with the replacement of chosenFemales/chosenMales with chosenAncestors

Commit:
Draft for table that display calculated COI and recommendation

Commit:
Table presentation of inbreeding calculation result

Commit:
renamed inbreeding to testbreed

Commit:
Made separate file for testbreed pedigree

Commit:
First draf tabbed inbreed recommendation where calculation and pedigree are separated into two tabs

HenrikeW committed:
created view for certificate-form

HenrikeW committed:
created view for certificate-form

Commit:
If both parents are registered, make use of the kinship matrix to speed things up

Commit:
First draft global Pedi variable

Commit:
return calculated COI with only two decimals

HenrikeW committed:
request to get father data from backend

HenrikeW committed:
request to get father data from backend

HenrikeW committed:
added mother and changed colour spelling in API

HenrikeW committed:
added mother and changed colour spelling in API

Commit:
fix update_data_cron now working

HenrikeW committed:
conditional rendering for summary

HenrikeW committed:
conditional rendering for summary

Commit:
Add method call that updates the data that the R-API uses when db herd related data is modified/added

HenrikeW committed:
added summary and authentification

HenrikeW committed:
added summary and authentification

pontus committed:
Bump versions of a lot of dependencies

pontus committed:
installOptions is now packageOptions

pontus committed:
Use npx to reference snowpack

pontus committed:
Run prepare on setup

HenrikeW committed:
small fixes for fetching ancestor data

HenrikeW committed:
small fixes for fetching ancestor data

jonandernovella committed:
Add deps to main container

jonandernovella committed:
Add support for issuing certificates

Certs are filled based on a pdf template and using pdfrw in order to write to the annotations.

HenrikeW committed:
removed isNew state

HenrikeW committed:
cleanup

HenrikeW committed:
more cleanup

HenrikeW committed:
more cleanup

HenrikeW committed:
Doing commit without message

Just trying things out.

jonandernovella committed:
Fix farfar key

HenrikeW committed:
Trying to fetch pdf from endpoint

Not done yet, just trying things with the GET request.

HenrikeW committed:
Trying to fetch pdf from endpoint

Not done yet, just trying things with the GET request.

Commit:
remove update R-api method from python api and add updata_data function in testbreed instead

Commit:
Merge branch 'develop' into feature/test-breed-functionality

pontus committed:
Create volume for node_modules and use it in the right place

pontus committed:
Bumped packages other than snowpack and date-io

Commit:
minor code formating

Commit:
minor code formatting

HenrikeW committed:
Fix get request to receive pdf file

Function issueCertificate fixed. User now able to open the pdf in a new tab.

HenrikeW committed:
Fix get request to receive pdf file

Function issueCertificate fixed. User now able to open the pdf in a new tab.

HenrikeW committed:
Installed npm package react-pdf

HenrikeW committed:
Installed npm package react-pdf

jonandernovella committed:
Merge pull request #217 from NBISweden/feature/code-ql-security-scanning

Use suggested action for codeql

Commit:
Merge pull request #234 from NBISweden/feature/issue-225-duplicates

Add test for duplicate numbers

andkaha committed:
Add missing arguments in example code

andkaha committed:
Update tracking years to go all the way to 2021 now

Also load body fat data for 2019 (was 2018 before).

HenrikeW committed:
Showing certificate as pdf

Using the react-pdf library to show an image of the certificate's
pdf-version within the page.

HenrikeW committed:
Prepare for preview endpoint

added a post request and a limited individual-object to make a request
to an endpoint that generates a preview pdf

HenrikeW committed:
removed useEffects for ancestor data

jonandernovella committed:
Add preview, issue and update endpoints

HenrikeW committed:
removed usestates for ancestor queries

jonandernovella committed:
Update db upon issue

HenrikeW committed:
changing preview-post request to open pdf

using ArrayBuffer to be able to save the pdf from the response of the POST request
and then open it with react-pdf library

jonandernovella committed:
Cleanup

jonandernovella committed:
Add more docstrings

jonandernovella committed:
Add static method for getting signer

jonandernovella committed:
Use static method in endpoint

HenrikeW committed:
Change issue-request to POST

Response is now saved as a blob and can be downloaded by the user.
For now still using the update-endpoint for convenience reasons.

andkaha committed:
This function is included in `healthcheck.sql`

andkaha committed:
This was needed after all

jonandernovella committed:
Add first s3 components

jonandernovella committed:
Send proper status codes

jonandernovella committed:
Remove static method ref

HenrikeW committed:
Error handling on the issue-cert request

Adding handling for the case when there already is digital certificate.
Also changing request to the issue endpoint.

andkaha committed:
Add number/year healthcheck

HenrikeW committed:
Cleanup

Removing the save-function and adding comments.

andkaha committed:
Fix syntax error, reflow comment

andkaha committed:
Comment fix ("first" --> "last")

Commit:
Merge pull request #253 from NBISweden/feature/load-2021

Load 2021 Gotland data, and "number" health-check.

jonandernovella committed:
Add more s3 stuff

jonandernovella committed:
More cleanup

pontus committed:
Remove any node_modules brought in from development environment

jonandernovella committed:
Add method for checksum verification

jonandernovella committed:
Simplify s3 setup

jonandernovella committed:
Add first verification components

jonandernovella committed:
Merge pull request #255 from NBISweden/feature/js_upgrade

Upgrade some js packages

HenrikeW committed:
Error handling in POST request & styling

Added cases for different status codes in issueCertificate.
Also did some styling for form and buttons.

jonandernovella committed:
Use logical folder

jonandernovella committed:
Use same s3 path for getting checksum

jonandernovella committed:
Improve error handling

jonandernovella committed:
Define s3 vars in settings

jonandernovella committed:
Set cert settings in a better place

jonandernovella committed:
Use the defined settings

jonandernovella committed:
Embed hostname in qr code

jonandernovella committed:
Do not use HOSTNAME env var

jonandernovella committed:
Clean up to make linter happier

jonandernovella committed:
Make the linter happy

jonandernovella committed:
Fix import

jonandernovella committed:
Add more docstrings

HenrikeW committed:
breeding, communication and data prettier

HenrikeW committed:
Explore, field, filter, genebank prettier

5 files

HenrikeW committed:
herd and individual files prettier

6 files

HenrikeW committed:
login, main, manage, message, nav., owner prettier

6 files

HenrikeW committed:
pedigree, sorted, ui prettier

4 files

HenrikeW committed:
user prettier

4 files

pontus committed:
Maintain cache of kinship values through apscheduler job

pontus committed:
Add function to get all genebanks

HenrikeW committed:
Merge branch 'develop' into feature/prettier

pontus committed:
Resolve jscpd linter complaint

pontus committed:
Resolve jscpd linter complaint

pontus committed:
Resolve jscpd linter complaints

pontus committed:
Resolve jscpd linter complaint

pontus committed:
Merge pull request #265 from NBISweden/feature/prettier

All files in /src formatted with Prettier extension

pontus committed:
Resolved conflicts

pontus committed:
Merge pull request #258 from NBISweden/feature/digital_certs_frontend

Functionality to issue a new digital certificate

pontus commented on app/herdbook.py:
```diff
@@ -497,13 +497,19 @@ def update_certificate(i_number):
+                pdf_bytes=signed_data.getvalue(), ind_number=ind["number"]
+            )
+        except Exception as ex:  # pylint: disable=broad-except
+            print(ex)
```
```suggestion
            APP.logger.info("Unexpected error while updating certificate "+str(ex))
```

jonandernovella replied:
Yes, that's good. For now I will make them all debug lines though.

pontus commented on app/herdbook.py:
```diff
@@ -523,20 +529,26 @@ def issue_certificate(i_number):
+        print(ex)
+        return jsonify({"response": "Error processing your request"}), 400
 
     if uploaded:
```
I'm not overly fond of this style because one needs to think about the flow to understand why this doesn't blow up with `uploaded` undefined. I much prefer either having the return within the try or making sure to define `uploaded` to `None` above the try.

jonandernovella replied:
This one is for capturing mostly S3 related failures, which are a bit tricky to propagate otherwise. I am happy to work on a better solution in the next iteration :)

pontus replied:
I think I may have been unclear, I'd be happy if there would be an `uploaded = False` above the code.

As far as I can see, the code above works (won't blow up on `uploaded` undefined), but it isn't obvious.

jonandernovella replied:
Aha! Alright now I follow you.

jonandernovella replied:
Should be fixed now

pontus commented on app/herdbook.py:
```diff
@@ -558,6 +570,38 @@ def preview_certificate(i_number):
+        signed = verify_signature(uploaded_bytes)
+        present = verify_certificate_checksum(ind["number"], checksum=checksum)
+    except Exception as ex:  # pylint: disable=broad-except
+        print(ex)
```
```suggestion
        APP.logger.info("Unexpected error while verifying certificate "+str(ex))
```

pontus commented on app/herdbook.py:
```diff
@@ -558,6 +570,38 @@ def preview_certificate(i_number):
+        print(ex)
+        return jsonify({"response": "Error processing your request"}), 400
+
+    if present and signed:
```
Same as above with preference for doing within try or defining before.

jonandernovella replied:
Done

pontus committed:
Fix logic for user selector

jonandernovella committed:
Make endpoint accept PATCH

jonandernovella committed:
Handle valid signature but not present file

jonandernovella committed:
Change url in qr code

Commit:
minor popup layout improvement

Commit:
prepare frontend for error handling

HenrikeW committed:
Merge pull request #267 from NBISweden/bugfix/user_editor_selector

Fix logic for user selector

Commit:
COI threshold depedent on genebank

HenrikeW committed:
Add update button and pass props

adding an additional button in individual_view and passing props
to individual_certificate to decide what action the component should be used for.

HenrikeW committed:
Change button text depending on action

Button fpr the last step in individual_certificate should change text depending on if the user is going to update or issue a certificate.

Commit:
update documentation

HenrikeW committed:
fix merge conflicts in individual_edit

HenrikeW committed:
add update-cert-functionality to frontend

changes in individual_certificate:
- added updateCertificate function
- adjusted JSX to show update/issue specific things

pontus committed:
Change colour to color, with migration

pontus committed:
Cache the correct function

pontus committed:
Linting fix

pontus committed:
Prettiered

pontus committed:
Merge branch 'develop' into fix/filter_active_or_dead

pontus committed:
Merge pull request #228 from NBISweden/fix/filter_active_or_dead

Update filter labels to match previously updated filter functionality

Commit:
Merge pull request #266 from NBISweden/feature/test-bread-faster

Add cache as suggested and start job refreshing cache at startup

jonandernovella committed:
Add missing dependencies to load script

jonandernovella committed:
Assign digital ids when issuing certs

The digital ids are present on the individual database for every entry, but they are only effective when assigned to the certificate attribute.

jonandernovella committed:
Create digital certificate id attribute

This id starts at 100000 and should be unique.

jonandernovella committed:
Use logger instead of print()

jonandernovella committed:
Log verification exception better

Co-authored-by: Pontus Freyhult <pontus_github@soua.net>

jonandernovella committed:
Log update exception better

Co-authored-by: Pontus Freyhult <pontus_github@soua.net>

jonandernovella committed:
Explicitely initialise boolean vars

Commit:
improve error handling

Commit:
Merge branch 'feature/test-breed-functionality' of github.com:NBISweden/theherdbook into feature/test-breed-functionality

Commit:
clean up

inghylt committed:
Merge branch 'develop' into feature/test-breed-functionality

Commit:
cleanup

jonandernovella committed:
Use request.json

jonandernovella committed:
Merge pull request #270 from NBISweden/bugfix/colour_is_color

Change colour to color, with migration

Commit:
Merge branch 'develop' into feature/test-breed-functionality

HenrikeW committed:
adding content-type to preview-request

adding "Content-Type": "application/json" to the previewCertificate-
function fixes the bug in the preview. Now data that was inserted in the
form should does show up in the preview.

Commit:
only return calculated data and not input data, redundant to return the input data

Commit:
lint fix

HenrikeW committed:
improving issueCertificate

jonandernovella committed:
Rework error handling for updating certs

HenrikeW committed:
improve error handling

HenrikeW committed:
removing console.log

HenrikeW committed:
cleaning code

Commit:
Change recommendation for Gotlandskanin

under 4% should be the recommended Inbreeding value

Commit:
Apply suggestions from code review

Just some minor text changes

HenrikeW committed:
Making error message in issueCertificate more specific

Co-authored-by: Pontus Freyhult <pontus_github@soua.net>

HenrikeW committed:
removing redundant code from error handling in updateCertificate

HenrikeW committed:
make ternary operator more specific / use switch instead

pontus committed:
Use memoize rather than cache for function

Commit:
Merge pull request #277 from NBISweden/feature/cache_fix

Use memoize rather than cache for function

Commit:
Merge pull request #227 from NBISweden/feature/test-breed-functionality

Feature/test breed functionality

jonandernovella committed:
Move sequence creation to init

jonandernovella committed:
Allow nullable digital cert ids

jonandernovella committed:
Get post data uniformly

jonandernovella committed:
Allow get request for previews

jonandernovella committed:
Move get data to correct scope

jonandernovella committed:
Check for database type

Co-authored-by: Pontus Freyhult <pontus_github@soua.net>

pontus committed:
Colour is color now

Commit:
Merge pull request #281 from NBISweden/bugfix/colour_is_color_in_scripts

Colour is color now

jonandernovella committed:
Add migration from 4 to 5

This includes the changes for digital certificate ids.

jonandernovella committed:
Bump schema version

jonandernovella committed:
Make the sequence optional

pontus committed:
Create simpler with sqlite

jonandernovella committed:
Handle S3 related vars properly

jonandernovella committed:
Don't hardcode template file

jonandernovella committed:
Add tests for certificate related endpoints

S3 mocking is done using moto.

jonandernovella committed:
Set up certs for testing

jonandernovella committed:
Make sure we update the db upon cert updates

jonandernovella committed:
Improve tests to match expectations

jonandernovella committed:
Use consistent name for colour refs

jonandernovella committed:
Test verify using unsigned certs

jonandernovella committed:
Make black happy

jonandernovella committed:
Remove quotation marks from env vars

jonandernovella committed:
Try to make python linters happy

HenrikeW committed:
creating certificate_form.tsx

copying JSX from individual_certificate, memoving buttons and adding basic imports

HenrikeW committed:
import CertificateForm with props

adding it to individual_certificate and passing style and individual

HenrikeW committed:
moving dropdown options to certificate_form

options for colors, sex and photos

HenrikeW committed:
note bug about canManage in individual_certificate

HenrikeW committed:
pass onUpdateIndividual as a prop

using handlers and prop drilling.
this way, the child component can update the state in the parent component

HenrikeW committed:
remove JSX from individual_certificate

now everything is in the child component certificate_form

jonandernovella committed:
Add more fixes

jonandernovella committed:
Add more cleanup

pontus committed:
More linter happiness

pontus committed:
Fix some shellcheck complaints also

HenrikeW committed:
pass canManage as prop to child component

HenrikeW committed:
remove unused inputs from individual_certificate

HenrikeW committed:
remove more unused code from individual_certificate

HenrikeW committed:
move heading to individual_certificate to make certificate_form more reusable

HenrikeW committed:
adjust comment about permissions

jonandernovella committed:
Merge pull request #292 from NBISweden/feature/add_rabbits

move herd form into a seperate component

jonandernovella committed:
Fix import conflict

jonandernovella committed:
Remove host port forwarding for s3 container
